File: java/350_IntersectionOfTwoArraysII.java
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        int[] bigger;
        int[] smaller;
        if (nums1.length > nums2.length) {
            bigger = nums1; smaller = nums2;
        } else {
            bigger = nums2; smaller = nums1;
        }
        
        HashMap<Integer, Integer> numCounts = new HashMap<Integer, Integer>();
        for (int n : smaller) {
            numCounts.put(n, numCounts.getOrDefault(n, 0) + 1);
        }
        
        int k = 0;
        List<Integer> resList = new ArrayList<Integer>();
        for (int n : bigger) {
            if (numCounts.get(n) != null && numCounts.get(n) != 0) {
                bigger[k++] = n;
                numCounts.put(n, numCounts.get(n)-1);
            }
        }
        return Arrays.copyOfRange(bigger, 0, k);
    }
}

File: cpp/350_IntersectionOfTwoArraysII.cpp
class Solution {
public:
    unordered_map<int, int> smaller_counts;
    
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        vector<int> bigger;
        vector<int> smaller_list;
        if (nums1.size() > nums2.size()) {
            bigger = nums1, smaller_list = nums2;
        } else {
            bigger = nums2, smaller_list = nums1;
        }
        
        for (int n : smaller_list) {
            smaller_counts[n] = smaller_counts.find(n) != smaller_counts.end() ? smaller_counts[n]+1 : 1;
        }
        
        vector<int> res = {};
        for (int n : bigger) {
            if (smaller_counts.find(n) != smaller_counts.end() && smaller_counts[n] != 0) {
                res.push_back(n);
                --smaller_counts[n];
            }
        }
        return res;
    }
};
